lib/spack/spack/database.py

```diff
@@ -372,9 +372,27 @@ def _read_suppress_error():
             old_data = self._data
             try:
                 self._data = {}
+                missing_specs = set(directory_layout.all_specs())
```

Comment:
why not just build up a set of added specs, and iterate over specs not in it later?  Seems clearer to me than discarding as each spec is installed.